modules/miner/update.go

```diff
@@ -1,11 +1,205 @@
 
 import (
-	"github.com/NebulousLabs/Sia/encoding"
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/types"
 )
 
+func (m *Miner) getNewSplitSets(diff *modules.TransactionPoolDiff) []*mapElement {
+	// Split the new sets and add the splits to the list of transactions we pull
+	// form.
+	newElements := make([]*mapElement, 0)
+	for _, newSet := range diff.AppliedTransactions {
+		// Split the sets into smaller sets, and add them to the list of
+		// transactions the miner can draw from.
+		//
+		// TODO: Split the one set into a bunch of smaller sets using the cp4p
+		// splitter.
+		m.setCounter++
+		m.fullSets[newSet.ID] = []int{m.setCounter}
+		var size uint64
+		var totalFees types.Currency
+		for i := range newSet.IDs {
+			size += newSet.Sizes[i]
+			for _, fee := range newSet.Transactions[i].MinerFees {
+				totalFees = totalFees.Add(fee)
+			}
+		}
+		// We will check to see if this splitSet belongs in the block.
+		s := &splitSet{
+			size:         size,
+			averageFee:   totalFees.Div64(size),
+			transactions: newSet.Transactions,
+		}
+
+		elem := &mapElement{
+			set:   s,
+			id:    splitSetID(m.setCounter),
+			index: 0,
+		}
+
+		m.splitSets[splitSetID(m.setCounter)] = s
+		newElements = append(newElements, elem)
+	}
+	return newElements
+}
+
+// addNewTxns adds new unconfirmed transactions to the miner's transaction
+// selection.
+func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	// Get new splitSets (in form of mapElement)
+	newElements := m.getNewSplitSets(diff)
+
+	// Place each elem in one of the MapHeaps
+	for _, elem := range newElements {
+		m.addMapElementTxns(elem)
+	}
+}
+
+// addMapElementTxns places the splitSet from a mapElement into the correct
+// mapHeap.
+func (m *Miner) addMapElementTxns(elem *mapElement) {
+	candidateSet := elem.set
+
+	// Check if heap for highest fee transactions has space.
+	if m.blockMapHeap.size+candidateSet.size < types.BlockSizeLimit-5e3 {
+		m.blockMapHeap.Push(elem)
+		return
+	}
+
+	// The block heap doesn't have enough space for this transaction.
+	// Check if removing sets from the blockMapHeap will be worth it.
+	// bottomSets will hold  the lowest fee sets from the blockMapHeap
+	bottomSets := make([]*mapElement, 0)
+	var sizeOfBottomSets uint64
+	var averageFeeOfBottomSets types.Currency
+
+	// While the heap cannot fit this set s, and while the (weighted)
+	// average fee for the lowest sets from the block is less than
+	// the fee for the set s, continue removing from the heap.
+	for {
+		_, exists := m.blockMapHeap.Peek()
+		// If the blockMapHeap is empty, push all elements removed from it
+		// back in, and place the candidate set into the overflow.
+		if !exists {
+			m.overflowMapHeap.Push(elem)
+			// Put back in transactions removed.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Go to the next candidate set.
+			break
+		}
+		nextSet := m.blockMapHeap.Pop()
+
+		bottomSets = append(bottomSets, nextSet)
```

Comment:
seems like it's one idea to fetch the next set and add it to the bottom set.

Also really seems like one idea to check if there is a set, then grab it, then add it to the bottom set. So this can probably all be one block of code.